src/writer.rs

```diff
@@ -149,51 +212,61 @@ where
 /// Write the data in 'f' order.
 ///
 /// Like NiBabel, we iterate by "slice" to improve speed and use less memory.
-fn write_data<A, S, D, W>(writer: &mut W, header: NiftiHeader, data: &ArrayBase<S, D>) -> Result<()>
+fn write_data<T, D, W>(writer: &mut W, header: NiftiHeader, data: ArrayView<T, D>) -> Result<()>
 where
-    S: Data<Elem = A>,
-    A: Copy,
-    A: DataElement,
-    A: Div<Output = A>,
-    A: FromPrimitive,
-    A: PodTransmutable,
-    A: ScalarOperand,
-    A: Sub<Output = A>,
+    T: Clone + PodTransmutable,
+    T: Div<Output = T>,
+    T: FromPrimitive,
+    T: PodTransmutable,
+    T: ScalarOperand,
+    T: Sub<Output = T>,
     D: Dimension + RemoveAxis,
     W: WriteBytesExt,
 {
-    // Need the transpose for fortran ordering used in nifti file format.
-    let data = data.t();
-
-    let mut write_slice = |data: Array<A, D::Smaller>| -> Result<()> {
-        let len = data.len();
-        let arr_data = data.into_shape(len).unwrap();
-        let slice = arr_data.as_slice().unwrap();
-        writer.write_all(guarded_transmute_to_bytes_pod_many(slice))?;
-        Ok(())
-    };
-
     // `1.0x + 0.0` would give the same results, but we avoid a lot of divisions
     let slope = if header.scl_slope == 0.0 {
         1.0
     } else {
         header.scl_slope
     };
     if slope != 1.0 || header.scl_inter != 0.0 {
-        let slope = A::from_f32(slope).unwrap();
-        let inter = A::from_f32(header.scl_inter).unwrap();
+        let slope = T::from_f32(slope).unwrap();
+        let inter = T::from_f32(header.scl_inter).unwrap();
         for arr_data in data.axis_iter(Axis(0)) {
-            write_slice(arr_data.sub(inter).div(slope))?;
+            write_slice(writer, arr_data.sub(inter).div(slope))?;
         }
     } else {
-        for arr_data in data.axis_iter(Axis(0)) {
-            // We need to own the data because of the into_shape() for `c` ordering.
-            write_slice(arr_data.to_owned())?;
-        }
+        write_slices(writer, data)?;
+    }
+    Ok(())
+}
+
+fn write_slices<T, D, W>(writer: &mut W, data: ArrayView<T, D>) -> Result<()>
+where
+    T: Clone + PodTransmutable,
+    D: Dimension + RemoveAxis,
+    W: WriteBytesExt,
+{
+    for arr_data in data.axis_iter(Axis(0)) {
+        // We need to own the data because of the into_shape() for `C` ordering.
+        write_slice(writer, arr_data.to_owned())?;
```

Comment:
Maybe we don't have to copy data here. Try passing an `ArrayView` instead and making `write_slice` accept any `ArrayBase<S, D>`.

Reply:
Without the copy the image will be flipped. The comment could probably be made clearer :) Both writing functions start with 

     // Need the transpose for fortran ordering used in nifti file format.
     let data = data.t();

but we know that `ndarray` doesn't actually flip the data, it simply changes the strides. The copy ensures that the data in memory is ordered so it can be fed to `guarded_transmute_to_bytes_pod_many`.

What I don't understand (a part of the code was made by a colleague who doesn't recall everything) is why we **always** need to flip. It makes sense only if `ndarray` doesn't actually care about 'c' or 'f' ordering and allocates your data in 'c' anyway. That's probably the case because our tests work!

Reply:
All right, this is interesting, to say the least. It seems reasonable for `to_owned` to just ignore the ordering of the original volume and arrange it in the default row-major ordering. In turn, the `into_shape` which flattens the array (line 264) will just discard the previously establish F order again.

Well, here's what we should do: we want to preserve memory usage while benefiting from fast writing of contiguous byte sections, so let's allocate a single array in C order and reuse it for every slice. Here's a patch with the trick:

```patch
diff --git a/src/writer.rs b/src/writer.rs
index 3a6cf46..f7a2b59 100644
--- a/src/writer.rs
+++ b/src/writer.rs
@@ -241,22 +241,31 @@ where
     Ok(())
 }
 
-fn write_slices<T, D, W>(writer: &mut W, data: ArrayView<T, D>) -> Result<()>
+fn write_slices<A, S, D, W>(writer: &mut W, data: ArrayBase<S, D>) -> Result<()>
 where
-    T: Clone + PodTransmutable,
+    S: Data<Elem = A>,
+    A: Clone + PodTransmutable,
     D: Dimension + RemoveAxis,
     W: WriteBytesExt,
 {
-    for arr_data in data.axis_iter(Axis(0)) {
-        // We need to own the data because of the into_shape() for `C` ordering.
-        write_slice(writer, arr_data.to_owned())?;
+    let mut iter = data.axis_iter(Axis(0));
+    if let Some(arr_data) = iter.next() {
+        // keep slice voxels in a separate array to ensure `C` ordering
+        // even after `into_shape`.
+        let mut slice = arr_data.to_owned();
+        write_slice(writer, slice.view())?;
+        for arr_data in iter {
+            slice.assign(&arr_data);
+            write_slice(writer, slice.view())?;
+        }
     }
     Ok(())
 }
 
-fn write_slice<T, D, W>(writer: &mut W, data: Array<T, D>) -> Result<()>
+fn write_slice<A, S, D, W>(writer: &mut W, data: ArrayBase<S, D>) -> Result<()>
 where
-    T: Clone + PodTransmutable,
+    S: Data<Elem = A>,
+    A: Clone + PodTransmutable,
     D: Dimension,
     W: WriteBytesExt,
 {
```